designs/2022-languages/README.md

```diff
@@ -0,0 +1,612 @@
+
+This proposal is 100% backwards compatible in the short-term, and all breaking changes are optional and can be phased in over time to avoid disrupting the ecosystem.
+
+This proposal does not require any changes from end-users.
+
+## Alternatives
+
+1. We could continue using the `parserForESLint()` approach to support other languages. This was proposed and accepted in [#56](https://github.com/eslint/rfcs/pull/56).
+1. We could choose to not support languages other than JavaScript.
+
+## Open Questions
+
+### Should `RuleContext` have mostly properties instead of a mix of properties and methods?
+
+With the proposal as-is, `RuleContext` will have four properties and five methods, but four of the five methods just return a value. If we are looking at a larger rewrite, should we create a cleaner interface with just properties? Such as:
+
+```ts
+interface RuleContext {
+
+    languageOptions: LanguageOptions;
+    settings: object;
+    options: Array<any>;
+    id: string;
+
+    // method replacements
+    cwd: string;
+    filename: string;
+    physicalFilename: string
+    sourceCode: SourceCode;
+
+    report(violation: Violation): void;
+
+}
+```
+
+### Should we eliminate `:exit` from selectors?
+
+The `:exit` was left over from before we used `esquery` and is a bit of an outlier from how selectors generally work. We could eliminate it and ask people to write their rule visitors like this:
+
+```js
+{
+    "FunctionExpression": {
+        enter(node, parent) {
+            // ...
+        },
+        exit(node, parent) {
+            // ...
+        }
+    }
+}
+```
+
+And if they just wanted to use `:exit`, we could ask them to write this:
+
+```js
+{
+    "FunctionExpression": {
+        exit(node, parent) {
+            // ...
+        }
+    }
+}
+```
+
+This would allow the visitor keys to stay strictly as selectors while still allowing both enter and exit phases of traversal to be hit.
+
+### Does the ESLint team want to create and maintain additional languages?
+
+The next logical step after this refactoring would be to create an additional language to ensure that other languages can be supported. JSON seems like a logical one (as many people seem to think ESLint validates JSON already).
+
+Does the ESLint team want to create an official JSON language plugin? Or any other language plugin? Or do we want to leave additional languages to the community?
+
+I think there is some benefit to having a few officially-supported language plugins that people will likely want, but that also brings additional maintenance burden.
+
+## Help Needed
+
+N/A
+
+## Frequently Asked Questions
+
+### Why allow binary files?
+
+I think there may be a use in validating non-text files such as images, videos, and WebAssembly files. I may be wrong, but by allowing languages to specify that they'd prefer to receive the file in binary, it at least opens up the possibility that people may create plugins that can validate binary files in some way.
```

Comment:
> In those situations, can we expect to have line and column location information? Some portions of the proposed API assume that. For binary files, we might only have an offset.

This is an excellent question that I haven't thought through. It is definitely possible that binary files will have no equivalent to line and column.

> Similarly, how much would allowing binary files alter fix/suggestion logic?

Because suggestions are mostly used by editors, and you probably aren't manually editing a binary file, I think that feature just wouldn't be used.

For fixing, that is an interesting question. It seems like we should be able to provide the ability to fix binary files in the same way we currently do text files, except the locations would be offsets and the data would always be numeric. So, we are probably looking at core `TextFixer` and `BinaryFixer` classes.

Reply:
Good point about suggestions.

Using offsets exclusively in binary files should work.

Reply:
Besides validating binary file, another benefits of allowing binary reading is using napi based native plugin. 
Language author can use napi to provide a interface to cooperate with eslint. The performance bottleneck is usually text-encoding, since most language uses utf8 instead of utf16 JavaScript uses. Allowing binary can enable us pass an ArrayBuffer to napi and reduce text-encoding overhead.